beacon-chain/state/setters.go

```diff
@@ -99,8 +102,16 @@ func (b *BeaconState) UpdateBlockRootAtIndex(idx uint64, blockRoot [32]byte) err
 		return fmt.Errorf("invalid index provided %d", idx)
 	}
 
-	// Copy on write since this is a shared array.
-	r := b.BlockRoots()
+	b.lock.RLock()
+	r := b.state.BlockRoots
+	if ref := b.sharedFieldReferences[blockRoots]; ref.refs > 1 {
+		ref.refs--
+		b.sharedFieldReferences[blockRoots] = &reference{refs: 1}
+
+		// Copy on write since this is a shared array.
+		r = b.BlockRoots()
```

Comment:
nitpick, would be great to follow the same ordering as the rest:
1.) Copy
2.) Decrease reference
3.) Set reference to 1

Reply:
Agreed, could be an issue!